### PR TITLE
Use the Epoch Manager contract to query for the latest epoch

### LINKED ISSUE #113: Wouldn't it be more concise if we compiled the contract only once?
Wouldn't it be more concise if we compiled the contract only once?

Maybe we can store it in the new `Oracle` struct and pass it as a parameter to this function.

Besides, if it were to fail, we would catch that right in the program start, rather than at the beginning of a new epoch.

Parsing the seems like a fast operation. I'm just wondering where the best place and moment to run this would be.

_Originally posted by @tilacog in https://github.com/edgeandnode/block-oracle/pull/111#discussion_r908918114_

### PR DESCRIPTION
**Pull request by tilacog**
Enables obtaining the latest epoch via a call to the Epoch Manager contract.

Fixes #113

Still needs to be tested before merging.

### Structural changes
- Created a `Contracts` object that will hold all logic for interacting with The Graph Protocol smart contracts.
- Placed it under the `Oracle` struct
- Moved the existing `DataEdge` interactions to the new `Contracts` struct.
- Config has now two contract addresses, named after DataEdge and Epoch Manager
- Removed and ignored `dev` config file, as it is often useful to use it locally with sensitive data __(private keys, URLs, etc.)__
- New `xtask` commands: `current-epoch` and `send-message`
- Added an `oracle/lib.rs` module to share some code with the `xtask` crate
- ~~Implemented backfilling~~

### Notes:
- Commit 564b44843943bb5951fac9e9b2736acb14584ded is the same as #143

### CODE REVIEW

**Comment by tilacog** [comment]
@neysofu, I'll revert the backfilling commits because:
- they currently work with an empty network list, and
- we might not need backfilling at all.

**Comment by tilacog** [comment]
Rebasing into the latest `main`.

File: crates/xtask/src/main.rs [hunk 1]
```diff
@@ -1,18 +1,76 @@
...
+            message,
+        } => contracts::send_message(message, environment).await?,
     };
     Ok(())
 }
+
+#[derive(clap::ValueEnum, Clone)]
+pub enum Message {
+    Reset,
+}
```
Comment (neysofu) [inline crates/xtask/src/main.rs:44]: This `enum` is effectively duplicating a subset of the `oracle-encoder` logic. We can get more features with less code if we simply ask the user for a `.json` file and compile that into a payload on the fly using `oracle-encoder` as a library:

`$ cargo xtask send-message ../config/dev/config.toml reset-message.json`

**Review by neysofu — APPROVED** [review]
I left a few comments. Most of these are not blockers, but rather ideas for improvements which can be addressed in future PRs. Feel free to fix as many of these as you want and drop the others.

File: crates/oracle/src/config.rs [hunk 2]
```diff
@@ -116,10 +118,11 @@ struct Clap {
 /// Represents the TOML config file
 #[derive(Deserialize)]
 #[serde(rename_all = "snake_case")]
-struct ConfigFile {
+pub struct ConfigFile {
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/oracle/src/config.rs:121]: Tracking this in #153. Thanks!

File: crates/oracle/src/contracts.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+use web3::{
+    api::Eth,
+    contract::{Contract, Options},
+    ethabi::Address,
+    types::{Bytes, H256, U256},
+    Transport,
+};
+
+const EPOCH_MANAGER_ABI: &[u8; 6563] = include_bytes!("abi/EpochManager.json");
+const DATA_EDGE_ABI: &[u8; 320] = include_bytes!("abi/DataEdge.json");
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/oracle/src/contracts.rs:13]: #154 :+1:

File: crates/oracle/src/subgraph.rs [hunk 4]
```diff
@@ -83,8 +83,7 @@ async fn query(url: Url) -> reqwest::Result<Response<graphql::subgraph_state::Re
         .unwrap();
     let request_body = graphql::SubgraphState::build_query(graphql::subgraph_state::Variables);
     let request = client.post(url).json(&request_body);
-    let response = request.send().await?;
-
+    let response = request.send().await?.error_for_status()?;
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/oracle/src/subgraph.rs:86]: I'll confirm if a failed subgraph returns a 200 or a 4XX/5XX.

File: .gitignore [hunk 5]
```diff
@@ -44,3 +44,4 @@ debug.log
 # oracle encoder json samples
 crates/oracle-encoder/examples/*.json
 !crates/oracle-encoder/examples/samples.json
+crates/oracle/config/dev/
```
[the remark this answers is not in the document]
Comment (tilacog) [reply .gitignore:47]: Let's revisit the configuration directory. We can keep a sample file and ignore dev/staging/prod files.

File: crates/oracle/config/prod/config.toml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+name = "eip155:1"
+jrpc = "http://example.com"
+polling_interval_in_seconds = 5
+
+[protocol_chain]
+name = "eip155:1"
+jrpc = "https://example.com"
+
+[indexed_chains]
+"spam:42" = "https://example.com"
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/oracle/config/prod/config.toml:18]: Mentioned here: https://github.com/graphprotocol/block-oracle/pull/135#discussion_r929903487

File: crates/oracle/config/test/config.sample.toml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+subgraph_url = "http://example.com"
+epoch_duration = 2
+
+[protocol_chain]
+name = "eip155:1"
+jrpc = "http://example.com"
+polling_interval_in_seconds = 5
+
+[indexed_chains]
+"spam:42" = "https://example.com"
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/oracle/config/test/config.sample.toml:15]: Mentioned here: https://github.com/graphprotocol/block-oracle/pull/135#discussion_r929903487

File: crates/oracle/src/oracle.rs [hunk 8]
```diff
@@ -110,7 +118,37 @@ impl Oracle {
...
 
-        Ok(self.epoch_tracker.is_new_epoch(block.number).await?)
+        Ok(self.is_new_epoch().await?)
+    }
+
+    pub async fn is_new_epoch(&self) -> Result<bool, Error> {
+        let subgraph_latest_epoch = match self
+            .subgraph_state
+            .last_state()
+            .ok_or(Error::MissingSubgraphState)?
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/oracle/src/oracle.rs]: Fixed in 50487e4

**Comment by neysofu** [comment]
@tilacog there's a conflict with `main` due to `Message::ChangeOwnership`, the fix is easy but let me know if you need help with it

File: crates/oracle/config/staging/config.toml [hunk 9]
```diff
@@ -1,19 +1,15 @@
...
 owner_private_key = "$PRIVATE_KEY"
-contract_address = "c4228701A7B92793b2428687DC37D0ed92e0d7A7"
-
+epoch_manager_address = "0x03541c5cd35953CD447261122F93A5E7b812D697"
+data_edge_address = "0xc4228701A7B92793b2428687DC37D0ed92e0d7A7"
 log_level = "trace"
 subgraph_url = "https://api.thegraph.com/subgraphs/name/juanmardefago/block-oracle"
-# FIXME: Once we have access to the Epoch Manager, we can completely remove this entry from configuration.
-# Should be about 10 minutes in Goerli
-epoch_duration = 40
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/oracle/config/staging/config.toml:9]: Fixed in 1068186

File: crates/oracle/src/subgraph.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/oracle/src/subgraph.rs:86]: It seems that it is standard GraphQL practice to separate HTTP status codes from errors (as GraphQL could be served in other transport layers). This means that the current code change will raise an error for 4XX/5XX HTTP status codes and not for subgraph errors, which are handled later in the code.

File: crates/xtask/src/main.rs [hunk 10]
```diff
@@ -1,18 +1,76 @@
...
+
+#[derive(clap::ValueEnum, Clone)]
+pub enum Environment {
+    Development,
+    Staging,
+    Production,
+}
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/xtask/src/main.rs]: Fixed in 87ba502

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (tilacog) [reply .gitignore]: Fixed in 9ca84f7

File: crates/xtask/src/main.rs (hunk 1 above)
Comment (tilacog) [reply crates/xtask/src/main.rs:50]: Referenced in a new issue: https://github.com/graphprotocol/block-oracle/issues/157

**Pull request merged**